---
title: "lab4_transcriptomics"
author: "Sarah Bowker"
date: "2025-09-25"
output: html_document
---

```{r upload_packages, message=FALSE, warning=FALSE}
BiocManager::install("clusterProfiler")
BiocManager::install("EnhancedVolcano")
#Note, if biomaRt isn't installed on your machine for some reason, you should install it as well. 

lapply(c(
  "docopt", "DT", "pheatmap","GenomicFeatures", "DESeq2",
  "edgeR", "systemPipeR", "systemPipeRdata", "BiocStyle", "GO.db", "dplyr",
  "tidyr", "stringr", "Rqc", "QuasR", "DT", "ape", "clusterProfiler","biomaRt","EnhancedVolcano"
),require, character.only = TRUE) 
```
# Q1: Change the path name and metadata table output

```{r, message=FALSE, warning=FALSE}
# Load required libraries
library(systemPipeRdata)
library(DT)

fq_path <- systemPipeRdata::pathList()$fastqdir

meta_data <- read.table(
  system.file("extdata", "param", "targetsPE.txt", package = "systemPipeRdata"),
  header = TRUE
)

head(meta_data)

meta_data$FileName1 <- file.path(fq_path, basename(meta_data$FileName1))
meta_data$FileName2 <- file.path(fq_path, basename(meta_data$FileName2))

DT::datatable(meta_data, options = list(pageLength = 10))
```

# Q2: Quality check

```{r quality_check, message=FALSE, warning=FALSE}

fastq_files <- c(meta_data$FileName1, meta_data$FileName2)
rqc_result <- Rqc::rqc(path = fq_path, pattern = ".fastq.gz")

rqcCycleQualityBoxPlot(rqc_result[1:12]) + theme_bw()
rqcCycleQualityBoxPlot(rqc_result[13:24]) + theme_bw()
rqcCycleQualityBoxPlot(rqc_result[25:36]) + theme_bw()

rqcCycleBaseCallsPlot(rqc_result[1:12]) + theme_bw()
rqcCycleBaseCallsPlot(rqc_result[13:24]) + theme_bw()
rqcCycleBaseCallsPlot(rqc_result[25:36]) + theme_bw()
```

# Q3. Process Reads
```{r process_reads, message=FALSE, warning=FALSE}
process_reads <- function(fwd_file_path) {
  fq <- readFastq(fwd_file_path)
  adapter_seq <- "GCCCGGGTAA"
  
  # Trim adapter sequence from 3' end
  fq_trimmed <- trimLRPatterns(Rpattern = adapter_seq, subject = fq)
  
  # Trim last 3 bases from 3' end
  width_trimmed <- width(fq_trimmed) - 3
  fq_trimmed <- narrow(fq_trimmed, start=1, end=width_trimmed)
  
  # Filter out reads with more than 1 'N'
  seqs <- sread(fq_trimmed)
  n_count <- vcountPattern("N", seqs)
  fq_filtered <- fq_trimmed[n_count <= 1]
  
  # Prepare output filename
  base_name <- tools::file_path_sans_ext(basename(fwd_file_path))
  out_file <- file.path("processed_reads", paste0(base_name, "_processed.fastq"))
  
  # Write filtered fastq
  writeFastq(fq_filtered, out_file, compress = FALSE, mode = "a")
  
  return(out_file)
}


if (!dir.exists("processed_reads")) dir.create("processed_reads")
for (i in seq_len(nrow(meta_data))) {
  old_path <- meta_data$FileName1[i]
  new_path <- process_reads(old_path)
  meta_data$FileName1[i] <- new_path  # Update metadata with processed file path
}
```